gefserver/pier/internal/dckr/dckr.go

```diff
@@ -525,48 +550,59 @@ func (c Client) StartExistingContainer(contID string, binds []string) (Container
 }
 
 // RemoveContainer removes a docker container
-func (c Client) RemoveContainer(containerID string) {
-	c.c.RemoveContainer(docker.RemoveContainerOptions{ID: containerID, Force: true})
+func (c Client) RemoveContainer(containerID string) error {
+	return c.c.RemoveContainer(docker.RemoveContainerOptions{ID: containerID, Force: true})
 }
 
-// WaitContainer takes a docker container and waits for its finish.
-// It returns the exit code of the container.
-func (c Client) WaitContainer(id ContainerID, removeOnExit bool) (int, error) {
+// WaitContainerOrSwarmService takes a docker container/swarm service and waits for its finish.
+// It returns the exit code of the container/swarm service.
+func (c Client) WaitContainerOrSwarmService(id string, removeOnExit bool) (int, error) {
+	// Inspection will fail, if the container does not exist
+	_, err := c.c.InspectContainer(id)
+	if err != nil {
+		return 0, nil
+	}
+
 	swarmOn, err := c.IsSwarmActive()
 	if err != nil {
 		return 1, err
 	}
 
+	// Swarm mode (swarm services)
 	if swarmOn {
-		notRunning := false
-		for {
-			notRunning, contState, err := c.IsSwarmContainerNotRunning(string(id))
+		isStopped, contState, err := c.isSwarmContainerStopped(id)
+		for isStopped == false {
+			isStopped, contState, err = c.isSwarmContainerStopped(id)
 			if (contState == swarm.TaskStateComplete || contState == swarm.TaskStateFailed || contState == swarm.TaskStateShutdown) && (err != nil) {
-				return 1, err
+				return 1, def.Err(err, "an error has occurred while executing a swarm service")
 			}
+		}
 
-			if notRunning {
-				break
-			}
+		serviceID, err := c.getSwarmServiceIDByContainerID(id)
+		if err != nil {
+			return 1, def.Err(err, "could not find a swarm service related to the provided container id: "+id)
 		}
 
-		if notRunning {
-			opts := docker.RemoveServiceOptions{ID: string(id)}
-			err = c.c.RemoveService(opts)
+		if removeOnExit {
+			if len(serviceID) > 0 {
```

Comment:
rewrite to `if serviceID != ""`, and can be combined with the previous condition: `if removeOnExit && serviceID != "" {`